database/src/ops/block.rs

```diff
@@ -1,89 +1,492 @@
 //! Blocks.
 
 //---------------------------------------------------------------------------------------------------- Import
+use std::sync::Arc;
 
-//---------------------------------------------------------------------------------------------------- Free Functions
-/// TODO
-pub fn add_block() {
-    todo!()
-}
+use curve25519_dalek::{constants::ED25519_BASEPOINT_POINT, Scalar};
+use monero_serai::{
+    block::Block,
+    transaction::{Input, Timelock, Transaction},
+};
 
-/// TODO
-pub fn add_block_data() {
-    todo!()
-}
+use cuprate_types::{ExtendedBlockHeader, TransactionVerificationData, VerifiedBlockInformation};
 
-/// TODO
-pub fn pop_block() {
-    todo!()
-}
+use crate::{
+    database::{DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    ops::{
+        blockchain::chain_height,
+        key_image::{add_key_image, remove_key_image},
+        macros::doc_error,
+        output::{
+            add_output, add_rct_output, get_rct_num_outputs, remove_output, remove_rct_output,
+        },
+        tx::{add_tx, get_num_tx, remove_tx},
+    },
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        AmountIndex, BlockHash, BlockHeight, BlockInfo, KeyImage, Output, OutputFlags,
+        PreRctOutputId, RctOutput, TxHash,
+    },
+    StorableVec,
+};
 
-/// TODO
-pub fn block_exists() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- `add_block_*`
+/// Add a [`VerifiedBlockInformation`] to the database.
+///
+/// This extracts all the data from the input block and
+/// maps/adds them to the appropriate database tables.
+///
+#[doc = doc_error!()]
+///
+/// # Panics
+/// This function will panic if `block.height > u32::MAX` (not normally possible).
+#[allow(clippy::too_many_lines)]
+// no inline, too big.
+pub fn add_block(
+    block: &VerifiedBlockInformation,
+    tables: &mut impl TablesMut,
+) -> Result<(), RuntimeError> {
+    // let new_block_height = chain_height(tables.block_heights())?;
+    // if block.height > new_block_height {
+    //     // ...provided a block that isn't (chain tip + 1)
+    // } else if block.height < new_block_height {
+    //     // ... provided a block that already exists
+    // }
```

Comment:
Should `ops/` functions react to user error?

Some conditions that aren't handled:
- Input block height > chain tip
- Input `{block,tx,output,etc}` already exists

What should happen here? Should we add/return a `RuntimeError::AlreadyExists` error?

Should `add_block()` and others just trust the caller and assume they satisfied documented preconditions?

Reply:
We could add asserts but I don't think these should be runtime errors

Reply:
We should trust the caller. We're most likely to have handled these input parameters higher in the code. therefore, checks should be done higher.

Reply:
Okay, adding asserts and panic docs for:
- `block.height > u32::MAX`
- `block.height` is not != `chain_height`

As for data that already exists, the function will just overwrite previous data.

Reply:
Sent message before saw yours @SyntheticBird45 :)

I'm ok with either but I think asserts further up the stack will just be optimized away if they happened in a previous function so more asserts is never bad?

(also I already wrote the tests...)

Reply:
the asserts may mess with testing btw

Reply:
Which testing? Should I remove them for now?

Reply:
> I think asserts further up the stack will just be optimized away if they happened in a previous function so more asserts is never bad?

You're right. asserts help rustc understand the bounds. I would say to got for it. Tho I don't see what @Boog900 is referring to

Reply:
If you have a test adding blocks to the DB with the assert that block height == chain height all test blocks will need to have their height as 0/ all blocks before it would need to be added. Not a big deal but still

Reply:
Ah, right. `block.height` is currently mutated in tests to avoid this.

I'll see if they have to be removed when we have real data.